### PR TITLE
[Core] Support Lark grammars for XGrammar

### PR DESCRIPTION
**Pull request by mgoin**
Expands XGrammar to support "all" types of grammars that outlines supports so we can remove the fallback

### CODE REVIEW

File: vllm/model_executor/guided_decoding/xgrammar_decoding.py [hunk 1]
```diff
@@ -152,7 +155,12 @@ def from_guided_params(cls,
                        tokenizer_hash=tokenizer_hash,
                        max_threads=max_threads)
         elif guided_params.grammar:
-            return cls(grammar_str=guided_params.grammar,
+            # XGrammar only supports GBNF grammars, so we must convert Lark
+            if grammar_is_likely_lark(guided_params.grammar):
+                grammar_str = convert_lark_to_gbnf(guided_params.grammar)
```
Comment (simon-mo) [inline vllm/model_executor/guided_decoding/xgrammar_decoding.py:160]: do a try catch here we can raise better error to the user something like "Failed to convert the grammar from Lark to GBNF, please use GBNF grammar or specify --guided-decoding-backend=outlines"

File: vllm/model_executor/guided_decoding/xgrammar_utils.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+        # Look for Lark-specific features
+        if any(pattern in line for pattern in ['?start:', '|', '~']):
+            return True
+
+    return False
+
+
+def convert_lark_to_gbnf(grammar_str: str) -> str:
+    """
```
Comment (simon-mo) [inline vllm/model_executor/guided_decoding/xgrammar_utils.py:41]: any sources or references that we can cite on this?

**Review by simon-mo — APPROVED** [review]

File: vllm/model_executor/guided_decoding/xgrammar_utils.py (hunk 2 above)
Comment (mgoin) [reply vllm/model_executor/guided_decoding/xgrammar_utils.py:41]: Unfortunately I couldn't find an existing converter, so I made this myself (with the help of AI) using the [GBNF reference](https://github.com/ggerganov/llama.cpp/blob/master/grammars/README.md) and the [Lark grammar reference](https://lark-parser.readthedocs.io/en/latest/grammar.html), so I can leave these links.

I tested that it works for our SQL-based grammars in examples and CI, but I wouldn't trust it blindly.

File: vllm/model_executor/guided_decoding/xgrammar_decoding.py (hunk 1 above)
Comment (mgoin) [reply vllm/model_executor/guided_decoding/xgrammar_decoding.py:160]: Yup, this is a good call

**Pull request merged**